File: backend/src/modulos/disponibilidad/disponibilidad.servicio.js
// HU-09 Gestionar disponibilidad (bloqueo de fechas)
const { consultar, transaccion } = require('../../db');
const { errores } = require('../../utils/errores');
const { hoy, buscarSuperposicion } = require('../../utils/fechas');

const ESTADOS_RESERVA_OCUPAN = ['PENDIENTE_PAGO', 'PAGADA'];

function mapearBloqueo(f) {
  return {
    id: f.id,
    fechaInicio: f.fecha_inicio,
    fechaFin: f.fecha_fin,
    motivo: f.motivo,
    creadoPor: f.creado_por_nombre ?? null,
    creadoEn: f.creado_en,
  };
}

async function asegurarMaquina(maquinaId, cliente = null) {
  const ejecutar = cliente ? cliente.query.bind(cliente) : consultar;
  const { rows } = await ejecutar(
    `SELECT id, estado FROM maquinas WHERE id = $1 ${cliente ? 'FOR UPDATE' : ''}`,
    [maquinaId]
  );
  if (!rows[0]) throw errores.noEncontrado('La máquina no existe');
  return rows[0];
}

async function listarBloqueos(maquinaId, { incluirPasados = false } = {}) {
  await asegurarMaquina(maquinaId);
  const { rows } = await consultar(
    `SELECT b.*, u.nombre AS creado_por_nombre
       FROM bloqueos_disponibilidad b
       LEFT JOIN usuarios u ON u.id = b.creado_por
      WHERE b.maquina_id = $1 AND ($2::boolean OR b.fecha_fin >= $3::date)
      ORDER BY b.fecha_inicio`,
    [maquinaId, incluirPasados, hoy()]
  );
  return rows.map(mapearBloqueo);
}

/**
 * Bloquea uno o varios rangos (HU-09 criterio 1) en una sola transacción:
 * o se crean todos, o ninguno.
 */
async function bloquear(maquinaId, { rangos, motivo }, usuarioId) {
  const fechaHoy = hoy();
  const pasado = rangos.find((r) => r.fechaInicio < fechaHoy);
  if (pasado) {
    throw errores.solicitudInvalida(`No se pueden bloquear fechas pasadas (${pasado.fechaInicio})`);
  }
  const cruce = buscarSuperposicion(rangos);
  if (cruce) {
    throw errores.solicitudInvalida(
      `Los rangos ${cruce[0].fechaInicio}–${cruce[0].fechaFin} y ${cruce[1].fechaInicio}–${cruce[1].fechaFin} se superponen entre sí`
    );
  }

  const creados = await transaccion(async (cliente) => {
    await asegurarMaquina(maquinaId, cliente);

    // HU-09 criterio 3: no se puede bloquear una fecha con una reserva ya pagada
    for (const r of rangos) {
      const { rows } = await cliente.query(
        `SELECT fecha_inicio, fecha_fin FROM reservas
          WHERE maquina_id = $1 AND estado = 'PAGADA'
            AND daterange(fecha_inicio, fecha_fin, '[]') && daterange($2::date, $3::date, '[]')
          ORDER BY fecha_inicio LIMIT 1`,
        [maquinaId, r.fechaInicio, r.fechaFin]
      );
      if (rows[0]) {
        throw errores.conflicto(
          `No se puede bloquear: hay una reserva pagada del ${rows[0].fecha_inicio} al ${rows[0].fecha_fin}`
        );
      }
    }

    const filas = [];
    for (const r of rangos) {
      const { rows } = await cliente.query(
        `INSERT INTO bloqueos_disponibilidad (maquina_id, fecha_inicio, fecha_fin, motivo, creado_por)
         VALUES ($1, $2, $3, $4, $5)
         RETURNING *`,
        [maquinaId, r.fechaInicio, r.fechaFin, motivo ?? null, usuarioId]
      );
      filas.push(rows[0]);
    }
    return filas;
  });

  return creados.map(mapearBloqueo);
}

/** HU-09 criterio 4: desbloquear fechas bloqueadas antes */
async function desbloquear(maquinaId, bloqueoId) {
  const { rowCount } = await consultar(
    'DELETE FROM bloqueos_disponibilidad WHERE id = $1 AND maquina_id = $2',
    [bloqueoId, maquinaId]
  );
  if (rowCount === 0) throw errores.noEncontrado('El bloqueo no existe');
}

/**
 * Fechas ocupadas de una máquina publicada (para el calendario público).
 * HU-09 criterio 5: los cambios se ven de inmediato porque se consulta en vivo.
 * No se expone el motivo del bloqueo ni quién reservó.
 */
async function ocupacionPublica(maquinaId, { desde, hasta }) {
  const maquina = await asegurarMaquina(maquinaId);
  if (maquina.estado !== 'PUBLICADA') {
    throw errores.noEncontrado('La máquina no existe o ya no está disponible');
  }
  const { rows } = await consultar(
    `SELECT fecha_inicio, fecha_fin, 'BLOQUEO' AS tipo
       FROM bloqueos_disponibilidad
      WHERE maquina_id = $1 AND daterange(fecha_inicio, fecha_fin, '[]') && daterange($2::date, $3::date, '[]')
     UNION ALL
     SELECT fecha_inicio, fecha_fin, 'RESERVA' AS tipo
       FROM reservas
      WHERE maquina_id = $1 AND estado = ANY($4::text[])
        AND daterange(fecha_inicio, fecha_fin, '[]') && daterange($2::date, $3::date, '[]')
     ORDER BY fecha_inicio`,
    [maquinaId, desde, hasta, ESTADOS_RESERVA_OCUPAN]
  );
  return {
    maquinaId,
    desde,
    hasta,
    ocupados: rows.map((f) => ({ fechaInicio: f.fecha_inicio, fechaFin: f.fecha_fin, tipo: f.tipo })),
  };
}

module.exports = { listarBloqueos, bloquear, desbloquear, ocupacionPublica };

File: backend/src/main/java/pe/upao/alquiler/disponibilidad/DisponibilidadServicio.java
package pe.upao.alquiler.disponibilidad;

import pe.upao.alquiler.bd.Bd;
import pe.upao.alquiler.bd.Fila;
import pe.upao.alquiler.config.Config;
import pe.upao.alquiler.json.Json;
import pe.upao.alquiler.maquinas.MaquinaRepositorio;
import pe.upao.alquiler.modelo.Bloqueo;
import pe.upao.alquiler.modelo.EstadoMaquina;
import pe.upao.alquiler.modelo.Maquina;
import pe.upao.alquiler.util.ErrorApp;
import pe.upao.alquiler.util.Fechas;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** HU-09 Gestionar disponibilidad (bloqueo de fechas) */
public class DisponibilidadServicio {

    private final Bd bd;
    private final BloqueoRepositorio bloqueos;
    private final MaquinaRepositorio maquinas;
    private final Config config;

    public DisponibilidadServicio(Bd bd, BloqueoRepositorio bloqueos, MaquinaRepositorio maquinas, Config config) {
        this.bd = bd;
        this.bloqueos = bloqueos;
        this.maquinas = maquinas;
        this.config = config;
    }

    private Maquina maquinaOFallar(long id) {
        Maquina m = maquinas.buscar(id);
        if (m == null) throw ErrorApp.noEncontrado("La máquina no existe");
        return m;
    }

    public List<Bloqueo> listar(long maquinaId, boolean incluirPasados) {
        maquinaOFallar(maquinaId);
        return bloqueos.listar(bd, maquinaId, incluirPasados, Fechas.hoy(config.zonaHoraria));
    }

    /**
     * Bloquea uno o varios rangos (HU-09 criterio 1) en una sola transacción:
     * o se crean todos, o ninguno.
     */
    public List<Bloqueo> bloquear(long maquinaId, List<Fechas.Rango> rangos, String motivo, long usuarioId) {
        String hoy = Fechas.hoy(config.zonaHoraria);
        for (Fechas.Rango r : rangos) {
            if (r.fechaInicio().compareTo(hoy) < 0) {
                throw ErrorApp.solicitudInvalida("No se pueden bloquear fechas pasadas (" + r.fechaInicio() + ")");
            }
        }
        Fechas.Rango[] cruce = Fechas.buscarSuperposicion(rangos);
        if (cruce != null) {
            throw ErrorApp.solicitudInvalida("Los rangos " + cruce[0].fechaInicio() + "–" + cruce[0].fechaFin() + " y "
                    + cruce[1].fechaInicio() + "–" + cruce[1].fechaFin() + " se superponen entre sí");
        }

        return bd.transaccion(tx -> {
            if (MaquinaRepositorio.bloquearFila(tx, maquinaId) == null) throw ErrorApp.noEncontrado("La máquina no existe");

            // HU-09 criterio 3: no se puede bloquear una fecha con una reserva ya pagada
            for (Fechas.Rango r : rangos) {
                Fila reserva = bloqueos.reservaPagadaQueSeCruza(tx, maquinaId, r.fechaInicio(), r.fechaFin());
                if (reserva != null) {
                    throw ErrorApp.conflicto("No se puede bloquear: hay una reserva pagada del "
                            + reserva.fecha("fecha_inicio") + " al " + reserva.fecha("fecha_fin"));
                }
            }
            List<Bloqueo> creados = new ArrayList<>();
            for (Fechas.Rango r : rangos) {
                creados.add(bloqueos.insertar(tx, maquinaId, r.fechaInicio(), r.fechaFin(), motivo, usuarioId));
            }
            return creados;
        });
    }

    /** HU-09 criterio 4 */
    public void desbloquear(long maquinaId, long bloqueoId) {
        if (!bloqueos.eliminar(bd, maquinaId, bloqueoId)) throw ErrorApp.noEncontrado("El bloqueo no existe");
    }

    /**
     * Fechas ocupadas de una máquina publicada (calendario público).
     * HU-09 criterio 5: se consulta en vivo, así que los cambios se ven de inmediato.
     * No se expone el motivo del bloqueo ni quién reservó.
     */
    public Map<String, Object> ocupacionPublica(long maquinaId, String desde, String hasta) {
        Maquina m = maquinas.buscar(maquinaId);
        if (m == null || m.estado() != EstadoMaquina.PUBLICADA) {
            throw ErrorApp.noEncontrado("La máquina no existe o ya no está disponible");
        }
        List<Map<String, Object>> ocupados = bloqueos.ocupacion(bd, maquinaId, desde, hasta).stream()
                .map(f -> Json.obj("fechaInicio", f.fecha("fecha_inicio"), "fechaFin", f.fecha("fecha_fin"), "tipo", f.texto("tipo")))
                .toList();
        return Json.obj("maquinaId", maquinaId, "desde", desde, "hasta", hasta, "ocupados", ocupados);
    }
}
